tls/s2n_record_write.c

```diff
@@ -157,20 +157,42 @@ S2N_RESULT s2n_record_min_write_payload_size(struct s2n_connection *conn, uint16
 int s2n_record_write_protocol_version(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
     uint8_t record_protocol_version = conn->actual_protocol_version;
+
+    /**
+     *= https://www.rfc-editor.org/rfc/rfc8446#section-5.1
+     *# This version value is historical, deriving from the use of 0x0301 for
+     *# TLS 1.0 and 0x0300 for SSL 3.0.  In order to maximize backward
+     *# compatibility, a record containing an initial ClientHello SHOULD have
+     *# version 0x0301 (reflecting TLS 1.0)
+     *
+     * We set actual_protocol_version early for clients, but we do not
+     * use that assumed value here in case we are talking to a legacy
+     * server that expects TLS1.0.
+     *
+     * If we are requesting early data, we can assume that we aren't talking to
+     * a legacy server as a legacy server would not know how to handle early data.
+     **/
     if (conn->server_protocol_version == s2n_unknown_protocol_version
             && conn->early_data_state != S2N_EARLY_DATA_REQUESTED) {
-        /* Some legacy TLS implementations can't handle records with protocol version higher than TLS1.0.
-         * To provide maximum compatibility, send record version as TLS1.0 if server protocol version isn't
-         * established yet, which happens only during ClientHello message. Note, this has no effect on
-         * protocol version in ClientHello, so we're still able to negotiate protocol versions above TLS1.0 */
         record_protocol_version = MIN(record_protocol_version, S2N_TLS10);
     }
 
-    /* In accordance to TLS 1.3 spec, https://tools.ietf.org/html/rfc8446#section-5.1
-     * tls record version should never be greater than 33 (legacy TLS 1.2 version).
-     */
+    /**
+     *= https://www.rfc-editor.org/rfc/rfc8446#section-5.1
+     *# legacy_record_version:  MUST be set to 0x0303 for all records
+     *#    generated by a TLS 1.3 implementation other than an initial
+     *#    ClientHello (i.e., one not generated after a HelloRetryRequest),
+     *#    where it MAY also be 0x0301 for compatibility purposes.
+     **/
     record_protocol_version = MIN(record_protocol_version, S2N_TLS12);
 
+    /* Never send an empty protocol version.
+     * If the protocol version is unknown, default to TLS1.0 like we do for initial ClientHellos.
+     */
+    if (record_protocol_version == s2n_unknown_protocol_version) {
+        record_protocol_version = S2N_TLS10;
```

Comment:
The suggestion in #1734 was to use 0x0303 which would be `S2N_TLS12`. Not sure if it matters though.

Reply:
I like TLS 1.0 here, since it seems like the recommendation in the RFC for providing backwards compatibility with TLS 1.0 servers should also apply to alerts when the version in unknown.

> In order to maximize backward
   compatibility, a record containing an initial ClientHello SHOULD have
   version 0x0301 (reflecting TLS 1.0)

Reply:
It's more compatible to use TLS1.0, as explained in the comments around what we do for initial ClientHellos.